src/data_race.rs

```diff
@@ -597,7 +597,7 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: MiriEvalContextExt<'mir, 'tcx> {
         // of the time, based on `rate`.
         let rate = this.machine.cmpxchg_weak_failure_rate;
         let cmpxchg_success = eq.to_scalar()?.to_bool()?
-            && (!can_fail_spuriously || this.machine.rng.get_mut().gen::<f64>() < rate);
+            && (!can_fail_spuriously || this.machine.rng.get_mut().gen::<f64>() >= rate);
```

Comment:
Please add a comment explaining `>=` vs `>`: `gen` returns a value from the half-open range `[0, 1)`, so `rate == 1` is guaranteed to lead to a `false` here.

Reply:
I think use of gen_bool makes it clear enough? Maybe?